mypy/build.py

```diff
@@ -861,59 +863,136 @@ def is_file(path: str) -> bool:
     return res
 
 
-def find_module(id: str, lib_path_arg: Iterable[str]) -> Optional[str]:
+USER_SITE_PACKAGES = \
+    'from __future__ import print_function; import site; print(site.getusersitepackages());' \
+    'print(*site.getsitepackages(), sep="\\n")'
+VIRTUALENV_SITE_PACKAGES = \
+    'from distutils.sysconfig import get_python_lib; print(get_python_lib())'
+
+
+def call_python(python_executable: str, command: str) -> str:
+    return subprocess.check_output([python_executable, '-c', command],
+                                   stderr=subprocess.PIPE).decode()
+
+
+@functools.lru_cache(maxsize=None)
+def get_site_packages_dirs(python_executable: str) -> List[str]:
+    """Find package directories for given python."""
+    if python_executable == sys.executable:
+        # Use running Python's package dirs
+        if hasattr(site, 'getusersitepackages') and hasattr(site, 'getsitepackages'):
+            user_dir = site.getusersitepackages()
+            return site.getsitepackages() + [user_dir]
+        # If site doesn't have get(user)sitepackages, we are running in a
+        # virtualenv, and should fall back to get_python_lib
+        return [get_python_lib()]
+    else:
+        # Use subprocess to get the package directory of given Python
+        # executable
+        try:
+            output = call_python(python_executable, USER_SITE_PACKAGES)
+        except subprocess.CalledProcessError:
+            # if no paths are found (raising a CalledProcessError), we fall back on sysconfig,
+            # the python executable is likely in a virtual environment, thus lacking
+            # needed site methods
+            output = call_python(python_executable, VIRTUALENV_SITE_PACKAGES)
+        return [line for line in output.splitlines() if os.path.isdir(line)]
+
+
+def find_base_dirs(lib_path: Tuple[str, ...], dir_chain: str, components: List[str],
+                   site_packages_dirs: List[str]) -> Tuple[Iterable[str], List[str]]:
+    # If we're looking for a module like 'foo.bar.baz', it's likely that most of the
+    # many elements of lib_path don't even have a subdirectory 'foo/bar'.  Discover
+    # that only once and cache it for when we look for modules like 'foo.bar.blah'
+    # that will require the same subdirectory.
+    dirs = find_module_dir_cache.get((dir_chain, lib_path), [])
+    if not dirs:
+        # Regular packages on the PATH
+        for pathitem in lib_path:
+            # e.g., '/usr/lib/python3.4/foo/bar'
+            isdir = find_module_isdir_cache.get((pathitem, dir_chain))
+            if isdir is None:
+                dir = os.path.normpath(os.path.join(pathitem, dir_chain))
+                isdir = os.path.isdir(dir)
+                find_module_isdir_cache[pathitem, dir_chain] = isdir
+            if isdir:
+                dirs.append(dir)
+        find_module_dir_cache[dir_chain, lib_path] = dirs
```

Comment:
@eric-wieser I realized I was essentially mutating the dir cache to add site packages then stripping that out. So I refactored things to not do that, which is much nicer.

Reply:
Much better!

Everything from here to the top of this function would be better as a separate helper function, because it would then be obvious that the caching is correct.

Reply:
Left a TODO.